src/services/anchor/index.ts

```diff
@@ -61,19 +64,71 @@ export const fetchTomlValues = async (TOML_FILE_URL: string): Promise<TomlValues
   };
 };
 
-async function getUrlParams(ephemeralAccount: string, supportsClientDomain: boolean): Promise<URLSearchParams> {
-  if (supportsClientDomain) {
-    return new URLSearchParams({ account: ephemeralAccount, client_domain: config.applicationClientDomain });
+//A memo derivation. TODO: Secure? how to check for collisions?
+async function deriveMemoFromAddress(address: `0x${string}`) {
+  const hash = keccak256(address);
+  return BigInt(hash).toString().slice(0, 15);
+}
+
+// Return the URLSearchParams and the account (master/omnibus or ephemeral) that was used for SEP-10
+async function getUrlParams(
+  ephemeralAccount: string,
+  usesMemo: boolean,
+  address: `0x${string}`,
+): Promise<{ urlParams: URLSearchParams; sep10Account: string }> {
+  if (usesMemo) {
+    const response = await fetch(`${SIGNING_SERVICE_URL}/v1/stellar/sep10`);
+
+    if (!response.ok) {
+      throw new Error('Failed to fetch client master SEP-10 public account.');
+    }
+    const { masterSep10Public } = await response.json();
+    if (!masterSep10Public) {
+      throw new Error('masterSep10Public not found in response.');
+    }
+
+    const sep10Account = masterSep10Public;
+
+    return {
+      urlParams: new URLSearchParams({
+        account: sep10Account,
+        client_domain: config.applicationClientDomain,
+        memo: await deriveMemoFromAddress(address),
+      }),
+      sep10Account,
+    };
   }
-  return new URLSearchParams({ account: ephemeralAccount });
+  return {
+    urlParams: new URLSearchParams({ account: ephemeralAccount, client_domain: config.applicationClientDomain }),
+    sep10Account: ephemeralAccount,
+  };
 }
 
+const sep10SignaturesWithLoginRefresh = async (
```

Comment:
Helper to retry once if the service returns a signature error, with a force refresh in the middle.

This is for the case the signer service has invalidated the signature, for some reason (it restarted or the user logged in elsewhere and then came back to use the old session)